KStreams shutdown - TimeoutException: The AdminClient thread is not accepting new calls

While upgrading to 1.7.0.Final I noticed a side-effect of #9967 in a corner case.

I believe this happens in a QuarkusTest without an embedded Kafka broker (so KafkaStreams is still in step waitForTopicsToBeCreated), which shuts down after JUnit test has completed.
The AdminClient is closed (https://github.com/quarkusio/quarkus/blob/master/extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsProducer.java#L113) which causes any call to the AdminClient be rejected with lots of these exceptions:

```
2020-08-13 21:14:05,523 ERROR [io.qua.kaf.str.run.KafkaStreamsProducer] (pool-24-thread-1)  Failed to get topic names from broker: java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.TimeoutException: Call(callName=listTopics, deadlineMs=1597346105523) timed out at 9223372036854775807 after 1 attempt(s)
        at org.apache.kafka.common.internals.KafkaFutureImpl.wrapAndThrow(KafkaFutureImpl.java:45)
        at org.apache.kafka.common.internals.KafkaFutureImpl.access$000(KafkaFutureImpl.java:32)
        at org.apache.kafka.common.internals.KafkaFutureImpl$SingleWaiter.await(KafkaFutureImpl.java:104)
        at org.apache.kafka.common.internals.KafkaFutureImpl.get(KafkaFutureImpl.java:272)
        at io.quarkus.kafka.streams.runtime.KafkaStreamsProducer.waitForTopicsToBeCreated(KafkaStreamsProducer.java:263)
        at io.quarkus.kafka.streams.runtime.KafkaStreamsProducer.access$000(KafkaStreamsProducer.java:51)
        at io.quarkus.kafka.streams.runtime.KafkaStreamsProducer$1.run(KafkaStreamsProducer.java:141)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.apache.kafka.common.errors.TimeoutException: Call(callName=listTopics, deadlineMs=1597346105523) timed out at 9223372036854775807 after 1 attempt(s)
Caused by: org.apache.kafka.common.errors.TimeoutException: The AdminClient thread is not accepting new calls.
```

As there is no sleep in the while loop in case of ExecutionException / TimeoutException this causes the CPU to go 100% and the JVM to eventually run out of memory.

This is not reproducible with https://github.com/quarkusio/quarkus-quickstarts/tree/master/kafka-streams-quickstart , despite having a similar test setup on my project.

--- a/extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsProducer.java
+++ b/extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsProducer.java
@@ -17,7 +17,7 @@ import java.util.concurrent.TimeoutException;
 import java.util.function.Function;
 import java.util.stream.Collectors;
 
-import javax.annotation.PreDestroy;
+import javax.enterprise.event.Observes;
 import javax.enterprise.inject.Instance;
 import javax.enterprise.inject.Produces;
 import javax.inject.Inject;
@@ -38,6 +38,7 @@ import org.apache.kafka.streams.processor.StateRestoreListener;
 import org.jboss.logging.Logger;
 
 import io.quarkus.arc.Unremovable;
+import io.quarkus.runtime.ShutdownEvent;
 import io.quarkus.runtime.Startup;
 
 /**
@@ -51,6 +52,7 @@ import io.quarkus.runtime.Startup;
 public class KafkaStreamsProducer {
 
     private static final Logger LOGGER = Logger.getLogger(KafkaStreamsProducer.class.getName());
+    private static volatile boolean shutdown = false;
 
     private final ExecutorService executorService;
     private final KafkaStreams kafkaStreams;
@@ -100,8 +102,8 @@ public class KafkaStreamsProducer {
         return kafkaStreamsTopologyManager;
     }
 
-    @PreDestroy
-    void close() {
+    void onStop(@Observes ShutdownEvent event) {
+        shutdown = true;
         if (executorService != null) {
             executorService.shutdown();
         }
@@ -143,8 +145,10 @@ public class KafkaStreamsProducer {
                     Thread.currentThread().interrupt();
                     return;
                 }
-                LOGGER.debug("Starting Kafka Streams pipeline");
-                kafkaStreams.start();
+                if (!shutdown) {
+                    LOGGER.debug("Starting Kafka Streams pipeline");
+                    kafkaStreams.start();
+                }
             }
         });
 
@@ -257,7 +261,7 @@ public class KafkaStreamsProducer {
     private static void waitForTopicsToBeCreated(Admin adminClient, Collection<String> topicsToAwait)
             throws InterruptedException {
         Set<String> lastMissingTopics = null;
-        while (true) {
+        while (!shutdown) {
             try {
                 ListTopicsResult topics = adminClient.listTopics();
                 Set<String> existingTopics = topics.names().get(10, TimeUnit.SECONDS);

--- a/integration-tests/kafka-streams/src/test/java/io/quarkus/it/kafka/streams/KafkaStreamsStartupFailureTest.java
+++ b/integration-tests/kafka-streams/src/test/java/io/quarkus/it/kafka/streams/KafkaStreamsStartupFailureTest.java
@@ -1,0 +1,62 @@
+package io.quarkus.it.kafka.streams;
+
+import static org.apache.kafka.streams.KafkaStreams.State;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.inject.Inject;
+
+import org.apache.http.HttpStatus;
+import org.apache.kafka.streams.KafkaStreams;
+import org.hamcrest.CoreMatchers;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Timeout;
+
+import io.quarkus.kafka.streams.runtime.KafkaStreamsTopologyManager;
+import io.quarkus.runtime.Application;
+import io.quarkus.test.common.QuarkusTestResource;
+import io.quarkus.test.junit.QuarkusTest;
+import io.quarkus.test.junit.QuarkusTestProfile;
+import io.quarkus.test.junit.TestProfile;
+import io.restassured.RestAssured;
+
+@QuarkusTestResource(KafkaTestResource.class)
+@TestProfile(KafkaStreamsStartupFailureTest.NonExistingTopicProfile.class)
+@QuarkusTest
+public class KafkaStreamsStartupFailureTest {
+
+    @Inject
+    KafkaStreams kafkaStreams;
+
+    @Inject
+    KafkaStreamsTopologyManager kafkaStreamsTopologyManager;
+
+    @Test
+    @Timeout(5)
+    public void testShutdownBeforeKStreamsStarted() throws Exception {
+        assertEquals(State.CREATED, kafkaStreams.state());
+        RestAssured.get("/health/ready").then()
+                .statusCode(HttpStatus.SC_SERVICE_UNAVAILABLE)
+                .body("checks[0].name", CoreMatchers.is("Kafka Streams topics health check"))
+                .body("checks[0].status", CoreMatchers.is("DOWN"))
+                .body("checks[0].data.missing_topics", CoreMatchers.is("nonexisting-topic"));
+        assertTrue(Application.currentApplication().isStarted());
+
+        Application.currentApplication().stop();
+
+        assertEquals(State.NOT_RUNNING, kafkaStreams.state());
+    }
+
+    public static class NonExistingTopicProfile implements QuarkusTestProfile {
+
+        @Override
+        public Map<String, String> getConfigOverrides() {
+            Map<String, String> conf = new HashMap<>();
+            conf.put("quarkus.kafka-streams.topics", "nonexisting-topic");
+            return conf;
+        }
+    }
+}

